pyquil/quil.py

```diff
@@ -209,13 +222,97 @@ def if_then(self, classical_reg, if_program, else_program=None):
         :returns: The Quil Program with the branching instructions added.
         :rtype: Program
         """
-
         else_program = else_program if else_program is not None else Program()
 
-        branch = If(Addr(classical_reg))
-        branch.Then.inst(if_program)
-        branch.Else.inst(else_program)
-        return self.inst(branch)
+        # IF [c]:
+        #    instrA...
+        # ELSE:
+        #    instrB...
+        #
+        # =>
+        #
+        #   JUMP-WHEN @THEN [c]
+        #   instrB...
+        #   JUMP @END
+        #   LABEL @THEN
+        #   instrA...
+        #   LABEL @END
+        label_then = LabelPlaceholder("THEN")
+        label_end = LabelPlaceholder("END")
+        self.inst(JumpWhen(target=label_then, condition=Addr(classical_reg)))
+        self.inst(else_program)
+        self.inst(Jump(label_end))
+        self.inst(JumpTarget(label_then))
+        self.inst(if_program)
+        self.inst(JumpTarget(label_end))
+        return self
+
+    def alloc(self):
+        """
+        Get a new qubit.
+
+        :return: A qubit.
+        :rtype: Qubit
+        """
+        return QubitPlaceholder()
+
+    def out(self):
+        """
+        Converts the Quil program to a readable string.
+
+        :return: String form of a program
+        :rtype: string
+        """
+        s = ""
+        for dg in self._defined_gates:
+            s += dg.out()
+            s += "\n"
+        for instr in self.instructions:
+            s += instr.out() + "\n"
+        return s
+
+    def get_qubits(self):
+        """
+        Returns all of the qubit indices used in this program, including gate applications and
+        allocated qubits. e.g.
+            >>> p = Program()
+            >>> p.inst(("H", 1))
+            >>> p.get_qubits()
+            {1}
```

Comment:
This should be `set([1])`, I think.

Reply:
I'm on Python 3 and the sets get printed with brackets

Reply:
Oof, of course!